client/client.go

```diff
@@ -300,17 +311,72 @@ func (client *Client) do(method, path string, query url.Values, headers map[stri
 		break
 	}
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer rsp.Body.Close()
+	return rsp, nil
+}
 
-	if v != nil {
-		if err := decodeInto(rsp.Body, v); err != nil {
-			return err
+// Do performs the HTTP request according to the provided options, possibly retrying GET requests
+// if appropriate for the status reported by the server.
+func (rq *defaultRequester) Do(ctx context.Context, opts *RequestOptions) (*RequestResponse, error) {
+	httpResp, err := rq.retry(ctx, opts.Method, opts.Path, opts.Query, opts.Headers, opts.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Is the result expecting a caller-managed raw body?
+	if opts.Type == RawRequest {
+		return &RequestResponse{Body: httpResp.Body}, nil
+	}
+
+	defer httpResp.Body.Close()
+	var serverResp response
+	if err := decodeInto(httpResp.Body, &serverResp); err != nil {
+		return nil, err
+	}
+
+	// Update the maintenance error state
+	if serverResp.Maintenance != nil {
+		rq.client.maintenance = serverResp.Maintenance
+	} else {
+		rq.client.maintenance = nil
+	}
+
+	// Deal with error type response
+	if err := serverResp.err(); err != nil {
+		return nil, err
+	}
+
+	// At this point only sync and async type requests may exist so lets
+	// make sure this is the case.
+	if opts.Type == SyncRequest {
```

Comment:
Nit: `switch opts.Type` with a `default` would be slightly nicer for this I think.

Reply:
Thanks, this does look much better.